vectordatasource/transform.py

```diff
@@ -8906,6 +8927,132 @@ def remap_viewpoint_kinds(shape, props, fid, zoom):
     return (shape, props, fid)
 
 
+def _list_of_countries(value):
+    """
+    Parses a comma or semicolon delimited list of ISO 3166-1 alpha-2 codes,
+    discarding those which don't match our expected format.
+
+    Returns a list of lower-case, stripped country codes.
+    """
+
+    from re import match
+    from re import split
+
+    countries = []
+    candidates = split('[,;]', value)
+
+    for candidate in candidates:
+        # should have an ISO 3166-1 alpha-2 code, so should be 2 ASCII
+        # latin characters.
+        candidate = candidate.strip().lower()
+        if match('[a-z][a-z]', candidate):
+            countries.append(candidate)
+
+    return countries
+
+
+def unpack_viewpoint_claims(shape, props, fid, zoom):
+    """
+    Unpack OSM "claimed_by" list into viewpoint kinds.
+
+    For example; "claimed_by=AA,BB,CC" should become "kind:aa=country,
+    kind:bb=country, kind:cc=country" (or region, etc... as appropriate for
+    the main kind, which should be "unrecognized_TYPE".
+    """
+
+    prefix = 'unrecognized_'
+    kind = props.get('kind')
+    claimed_by = props.get('claimed_by')
+
+    if kind and kind.startswith(prefix) and claimed_by:
+        claimed_kind = kind[len(prefix):]
+
+        for country in _list_of_countries(claimed_by):
+            props['kind:' + country] = claimed_kind
+
+    return (shape, props, fid)
+
+
+def apply_disputed_boundary_viewpoints(ctx):
+    """
+    """
+
+    from shapely.geometry import CAP_STYLE
+    from shapely.geometry import JOIN_STYLE
+
+    params = _Params(ctx, 'apply_disputed_boundary_viewpoints')
+    layer_name = params.required('base_layer')
+    start_zoom = params.optional('start_zoom', typ=int, default=0)
+    end_zoom = params.optional('end_zoom', typ=int)
+
+    layer = _find_layer(ctx.feature_layers, layer_name)
+    zoom = ctx.nominal_zoom
+
+    if zoom < start_zoom or \
+       (end_zoom is not None and zoom >= end_zoom):
+        return None
+
+    # we buffer out the shape by a small amount so that we're more likely to
+    # get a clean cut against the boundary line. with lines intersecting
+    # against lines, we often see a sort of "dashed pattern" where numerical
+    # imprecision has meant two lines don't quite intersect.
+    #
+    # tolerance for zoom is the length of 1px at 256px per tile, so we can take
+    # a fraction of that to get sub-pixel alignment.
+    buffer_distance = 0.1 * tolerance_for_zoom(zoom)
+
+    # first, separate out the dispute mask geometries
+    masks = []
+    other_features = []
```

Comment:
`uncontested_features`? By the time I get to `if not masks:` below I'm not sure what `other_features` means, the name change would clarify that.

Reply:
or `default_features` or? Since we're then applying the masks to the "other" features.

Reply:
I've refactored the whole thing in ed6f1363 and fc2a7ad6. Hopefully it's clearer now. I ended up splitting the features into `masks`, `boundaries` and putting anything non-matching straight into `new_features` to be returned, so the variable names should match up with their contents.